telemtry.md

```diff
@@ -0,0 +1,132 @@
+# Bluesky Enhancement Proposal: Telemetry
+
+## Abstract
+
+The goal is to capture the timing of every asynchronous action—meaning that the action that is done through a Status object—along with the context necessary to interpret that action, such as initial and final position of a movement or acquire time for an exposure. The set of asynchronous actions currently includes trigger, set, kickoff, complete. It should soon include stage and unstage as well when they operate via Status objects (though that work is not in scope for this proposal).
+
+This will enable future work on *time estimation* for bluesky plans. It is a necessary but not sufficient step toward that feature. Time estimation will additionally need to account for the time spent doing many fast operations, such as read() and describe(). It was proposed to record telemetry for these as well, but that proposal was rejected because recording the telemetry takes time comparable to the time of the action itself.
+
+## Affected Repositories
+
+- ophyd
+
+## Proposed Design
+
+Key aspects
+
+- Global function for adding telemetry destinations (can be 0, 1, or N) — could be local file, MongoDB, Elastic, etc.
+- Global function for inserting a new telemetry payload
+- The telemetry payloads are written and submitted from inside Status objects
+- No changes needed to OphydObj
+
+```python
+# New module-level functions in ophyd
+
+telemetry_destinations = []
+
+def register_telemetry(func):
+    telemetry_destinations.append(func)
+
+def record_telemetry(info):
+    for destination in telemetry_destinations:
+        destination(info)
+
+#  Update existing status objects
+
+class DeviceStatus:
+    ...
+    def __init__(self, device, **kwargs, command=None):
+        self.device = device
+        self._watchers = []
+        self.command = command 
+        # if command is None no telemetry is stored, else it must be
+        # 'set' or 'trigger' or ....
+        super().__init__(**kwargs)
+    
+    device_components = ()
+    status_attributes = ('start_time')
+
+    ...
+  
+    def _finished(self):
+        if self.command:  # For backwards compatibility
+            configuration = {'device_name': self.device.name,
+                             'command': status.command, 
+                             'device_components':{},
+                             'status_attributes':{}}
+            telemetry_dict = {}
+            for component_name in self.device_components:
+                telemetry_dict['device_components'][component_name] = getattr(self.device, component_name).get()
+            for attribute_name in self.status_attributes:
+                telemetry_dict['status_attributes'][attribute_name] = getattr(self, attribute_name).get()
+            telemetry_dict['status_attributes']['stop_time'] = time.time()
```

Comment:
To simplify you could set `self.stop_time = time.time()` somewhere above here, and then just include `'stop_time'` in `status_attributes` with everything else.

Reply:
I agree, but have done it this way to ensure that the `'stop_time'` is as close to insertion as possible. I am happy to change this if you prefer.

Reply:
Interesting question: do we include telemetry collection in the measurement?

Reply:
From the point of view of tracking hardware performance over time, we should measure stop time as early as possible here and not include the overhead of the get() calls that collect telemetry info. From the point of view of accurate time estimation, we will need to account for that overhead somehow.

Perhaps we should record stop_time of the action itself (measured first thing in this method) and then also record “overhead” in seconds as a separate key.

Reply:
Ok I agree that we should probably record both, it also allows us to track the overhead of telemetry recording as well which could be quite useful. I will make the change tomorrow (unless you want too).

I am thinking “ insertion_time” for the third time, giving start_time, stop_time and insertion_time as the three time related inputs.

Reply:
I would call the time it takes to insert the `telemetry_dict` into telemetry database(s) "insertion time", and I think it would be confusing to call whatever this is "insertion time".

I think on our call you floated the idea of storing the absolute start time and then the other times as deltas. Do you think that's a good idea? As I think about it, I expect we will most often want to manipulate these as deltas.

Can you add a paragraph to the Abstract explaining the different times and what they measure?

Reply:
I am happy to store deltas instead, I assume then that “start_time” is the time at the start of the action. I would recommend “command_time” and “database_overhead” as the two different deltas, what do others think?

Reply:
Once we decide on the names I will add the time descriptions to the abstract.

Reply:
Hmm, I’m still not sure we are talking about the same thing. How could we include the database overhead? We won’t know it until after the data has already been inserted.

Reply:
Hmm, I’m still not sure we are talking about the same thing. How could we include the database overhead? We won’t know it until after the data has already been inserted.

Reply:
Let's take a step back. In fe09309b3e2530a50f76c3bc5a0f3e0e64bfa7b1 I spelled out the timing sequence. Let's continue this discussion with reference to that, in the main comments.